src/haven/devices/detectors/dxp.py

```diff
@@ -0,0 +1,233 @@
+# )
+
+
+class NDFileNetCDFIO(NDPluginBaseIO):
+    def __init__(self, prefix: str, name="") -> None:
+        # Define some signals
+        self.file_path = epics_signal_rw_rbv(str, prefix + "FilePath")
+        self.file_name = epics_signal_rw_rbv(str, prefix + "FileName")
+        self.file_path_exists = epics_signal_r(bool, prefix + "FilePathExists_RBV")
+        self.file_template = epics_signal_rw_rbv(str, prefix + "FileTemplate")
+        self.full_file_name = epics_signal_r(str, prefix + "FullFileName_RBV")
+        self.file_write_mode = epics_signal_rw_rbv(
+            adcore.FileWriteMode, prefix + "FileWriteMode"
+        )
+        self.num_capture = epics_signal_rw_rbv(int, prefix + "NumCapture")
+        self.num_captured = epics_signal_r(int, prefix + "NumCaptured_RBV")
+        self.lazy_open = epics_signal_rw_rbv(bool, prefix + "LazyOpen")
+        self.capture = epics_signal_rw_rbv(bool, prefix + "Capture")
+        self.array_size0 = epics_signal_r(int, prefix + "ArraySize0")
+        self.array_size1 = epics_signal_r(int, prefix + "ArraySize1")
+        self.create_directory = epics_signal_rw_rbv(int, prefix + "CreateDirectory")
+        super().__init__(prefix, name)
+
+
+class NetCDFWriter(DetectorWriter):
+    def __init__(
+        self,
+        netcdf: NDFileNetCDFIO,
+        path_provider: PathProvider,
+        name_provider: NameProvider,
+        dataset_describer: DatasetDescriber,
+        *plugins: adcore.NDArrayBaseIO,
+    ) -> None:
+        self.netcdf = netcdf
+        self._path_provider = path_provider
+        self._name_provider = name_provider
+        self._dataset_describer = dataset_describer
+
+        self._plugins = plugins
+        self._capture_status: AsyncStatus | None = None
+
+    async def open(self, multiplier: int = 1) -> dict[str, DataKey]:
+        raise NotImplementedError()
+
+    async def close(self):
+        raise NotImplementedError()
+
+    async def observe_indices_written(
+        self, timeout=DEFAULT_TIMEOUT
+    ) -> AsyncGenerator[int, None]:
+        raise NotImplementedError()
+
+    async def get_indices_written(self) -> int:
+        raise NotImplementedError()
+
+    async def collect_stream_docs(
+        self, indices_written: int
+    ) -> AsyncIterator[StreamAsset]:
+        raise NotImplementedError()
+
+
+class ScanRate:
+    pass
+
+
+class DXPDriverIO(Device):
+
+    class PresetMode(StrictEnum):
+        NO_PRESET = "No preset"
+        REAL_TIME = "Real time"
+        LIVE_TIME = "Live time"
+        EVENTS = "Events"
+        TRIGGERS = "Triggers"
+
+    class CollectMode(StrictEnum):
+        MCA_SPECTRA = "MCA spectra"
+        MCA_MAPPING = "MCA mapping"
+        SCA_MAPPING = "SCA mapping"
+        LIST_MAPPING = "List mapping"
+
+    def __init__(self, prefix: str, name: str = "") -> None:
+        # SNL status records
+        self.snl_connected = epics_signal_r(bool, f"{prefix}SNL_Connected")
+        # Acquisition control records
```

Comment:
I vaguely remebered there is also one more option erase_start in epics, but maybe it's not necessary to include this

Reply:
Yep, there is.

For the xspress detectors I found that if you start the HDF writer, and then use "erase", you get an extra frame in the HDF file because "erase" sets the ndarray to all zeros and the writer treats that as new data to save.

So to avoid this extra frame, the Xspress device sets "erase", then starts the HDF plugins, then sets "start". I couldn't think of a situation where "start_erase" was necessary, I left it off of the xspress device so people don't accidentally save extra blank frames.

I'm not sure if that extra frame happens on the DXP detectors, though, so if there's a reason to include it we maybe could. Do you have a use case in mind?

Reply:
Let's just leave if off